File: python/tests/fake_daemon.py
#!/usr/bin/env python3
"""A stand-in for the Go transport, so the client's protocol handling can be
tested without Go and without a network.

Every behaviour the real transport can exhibit — including the ones that only
happen when something has gone wrong — is reachable by asking for it in the URL.
That is the point: the interesting code in the client is what it does with a
late answer, a garbage line or a dead pipe, and none of those are reproducible
against a healthy server.
"""

from __future__ import annotations

import base64
import json
import sys
import threading
import time
from urllib.parse import urlsplit


def emit(obj: dict) -> None:
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()


def answer(request: dict, **extra: object) -> dict:
    return {
        "id": request.get("id"),
        "status": 200,
        "url": request.get("url", ""),
        # The body echoes the request, so a test can assert what actually
        # reached the transport rather than what the client meant to send.
        "body": json.dumps(
            {
                "argv": sys.argv[1:],
                "method": request.get("method"),
                "headers": request.get("headers", {}),
                "order": request.get("order", []),
                "cookies": request.get("setCookie", []),
                "body": request.get("body"),
            }
        ),
        "headers": {
            "content-type": ["application/json"],
            "set-cookie": ["a=1", "b=2"],
        },
        "cookies": [],
        **extra,
    }


def main() -> None:
    for line in sys.stdin:
        try:
            request = json.loads(line)
        except ValueError:
            emit({"id": 0, "error": "bad request"})
            continue

        behaviour = urlsplit(request["url"]).hostname

        if behaviour == "ok":
            emit(answer(request))
        elif behaviour == "error":
            emit({"id": request["id"], "error": "upstream refused"})
        elif behaviour == "null-fields":
            # What the transport really sends when a request fails: the shape is
            # whole, the fields are null.
            emit({"id": request["id"], "status": 0, "url": "", "body": "",
                  "headers": None, "cookies": None})
        elif behaviour == "bad-error":
            emit(answer(request, error={}))
        elif behaviour == "bad-status":
            emit(answer(request, status="200"))
        elif behaviour == "bad-url":
            emit(answer(request, url=7))
        elif behaviour == "bad-body":
            emit(answer(request, body=[]))
        elif behaviour == "binary":
            # The bytes a JSON string cannot hold: PNG magic and a stray 0xff.
            raw = bytes([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0xFF])
            emit(
                answer(
                    request,
                    body=base64.b64encode(raw).decode("ascii"),
                    bodyEncoding="base64",
                )
            )
        elif behaviour == "echo-body":
            # Hands back what it was sent, so the request direction can be checked.
            emit(
                answer(
                    request,
                    body=request.get("body", ""),
                    bodyEncoding=request.get("bodyEncoding", ""),
                )
            )
        elif behaviour == "utf8-encoding":
            emit(answer(request, body="plain text", bodyEncoding="utf8"))
        elif behaviour == "bad-body-encoding":
            emit(answer(request, body="x", bodyEncoding="rot13"))
        elif behaviour == "bad-body-encoding-type":
            emit(answer(request, body="x", bodyEncoding=7))
        elif behaviour == "bad-base64":
            emit(answer(request, body="!!not base64!!", bodyEncoding="base64"))
        elif behaviour == "bad-headers":
            emit(answer(request, headers=[]))
        elif behaviour == "bad-header-scalar":
            emit(answer(request, headers={"broken": 7}))
        elif behaviour == "bad-header-list":
            emit(answer(request, headers={"broken": [7]}))
        elif behaviour == "bad-cookies":
            emit(answer(request, cookies={}))
        elif behaviour == "bad-cookie-item":
            emit(answer(request, cookies=[7]))
        elif behaviour == "garbage":
            sys.stdout.write("not json at all\n")
            sys.stdout.flush()
        elif behaviour == "null-line":
            sys.stdout.write("null\n")
            sys.stdout.flush()
        elif behaviour == "no-id":
            emit({"status": 200, "body": "anonymous"})
            threading.Timer(0.05, lambda: emit(answer(request))).start()
        elif behaviour == "wrong-id":
            emit({"id": request["id"] + 1000, "status": 200, "body": "stray"})
            # …then the right one, so the request still completes and the test
            # can assert the stray line was ignored rather than merely slow.
            threading.Timer(0.05, lambda: emit(answer(request))).start()
        elif behaviour == "silent":
            pass  # Never answers. The client's deadline is the only thing that ends this.
        elif behaviour == "slow":
            threading.Timer(0.4, lambda: emit(answer(request))).start()
        elif behaviour == "exit":
            sys.exit(3)
        elif behaviour == "stderr":
            sys.stderr.write("a note on stderr\n")
            sys.stderr.flush()
            time.sleep(0.05)
            emit(answer(request))
        elif behaviour == "legacy-headers":
            emit(answer(request, headers={"content-type": "application/json"}))
        else:
            emit(answer(request, status=404))


if __name__ == "__main__":
    main()

File: node/test/fake-daemon.js
#!/usr/bin/env node
// A stand-in for the Go transport, so the client's protocol handling can be
// tested without Go and without a network.
//
// Every behaviour the real transport can exhibit — including the ones that only
// happen when something has gone wrong — is reachable by asking for it in the
// URL. That is the point: the interesting code in the client is what it does
// with a late answer, a garbage line or a dead pipe, and none of those are
// reproducible against a healthy server.

import readline from 'node:readline';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const here = path.dirname(fileURLToPath(import.meta.url));

const reader = readline.createInterface({ input: process.stdin });

reader.on('line', (line) => {
  let request;
  try {
    request = JSON.parse(line);
  } catch {
    process.stdout.write('{"id":0,"error":"bad request"}\n');
    return;
  }

  const behaviour = new URL(request.url).host;
  const reply = (extra) =>
    process.stdout.write(
      JSON.stringify({
        id: request.id,
        status: 200,
        url: request.url,
        body: JSON.stringify({
          argv: process.argv.slice(2),
          headers: request.headers ?? {},
          order: request.order ?? [],
          cookies: request.setCookie ?? [],
        }),
        headers: { 'content-type': ['application/json'], 'set-cookie': ['a=1', 'b=2'] },
        cookies: [],
        ...extra,
      }) + '\n',
    );

  switch (behaviour) {
    case 'ok':
      reply();
      break;
    case 'error':
      process.stdout.write(JSON.stringify({ id: request.id, error: 'upstream refused' }) + '\n');
      break;
    case 'garbage':
      process.stdout.write('not json at all\n');
      break;
    case 'null-line':
      process.stdout.write('null\n');
      break;
    case 'wrong-id':
      process.stdout.write(JSON.stringify({ id: request.id + 1000, status: 200, body: 'stray' }) + '\n');
      // …then the right one, so the request still completes and the test can
      // assert the stray line was ignored rather than merely slow.
      setTimeout(() => reply(), 10);
      break;
    case 'silent':
      // Never answers. The client's deadline is the only thing that ends this.
      break;
    case 'exit':
      process.exit(3);
      break;
    case 'stderr':
      process.stderr.write('a note on stderr\n');
      reply();
      break;
    case 'stderr-chunks':
      process.stderr.write('one logical');
      setTimeout(() => {
        process.stderr.write(' line\nsecond line\n');
        reply();
      }, 5);
      break;
    case 'legacy-headers':
      reply({ headers: { 'content-type': 'application/json' } });
      break;
    case 'null-headers':
      reply({ headers: null });
      break;
    case 'null-fields':
      reply({ status: null, url: null, body: null, headers: null, cookies: null });
      break;
    case 'bad-error':
      reply({ error: {} });
      break;
    case 'bad-status':
      reply({ status: '200' });
      break;
    case 'bad-url':
      reply({ url: 7 });
      break;
    case 'bad-body':
      reply({ body: [] });
      break;
    case 'fixture': {
      // The committed wire format, replayed verbatim except for the id, which
      // the client matches on. See testdata/protocol/README.md.
      const fixture = JSON.parse(
        readFileSync(path.join(here, '..', '..', 'testdata', 'protocol', 'response.json'), 'utf8'),
      );
      process.stdout.write(JSON.stringify({ ...fixture, id: request.id }) + '\n');
      break;
    }
    case 'binary':
      // The bytes a JSON string cannot hold: PNG magic and a stray 0xff.
      reply({
        body: Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0xff]).toString('base64'),
        bodyEncoding: 'base64',
      });
      break;
    case 'echo-body':
      // Hands back what it was sent, so the request direction can be checked.
      reply({ body: request.body ?? '', bodyEncoding: request.bodyEncoding ?? '' });
      break;
    case 'utf8-encoding':
      reply({ body: 'plain text', bodyEncoding: 'utf8' });
      break;
    case 'bad-body-encoding':
      reply({ body: 'x', bodyEncoding: 'rot13' });
      break;
    case 'bad-body-encoding-type':
      reply({ body: 'x', bodyEncoding: 7 });
      break;
    case 'bad-base64':
      reply({ body: '!!not base64!!', bodyEncoding: 'base64' });
      break;
    case 'bad-headers':
      reply({ headers: [] });
      break;
    case 'bad-header-scalar':
      reply({ headers: { broken: 7 } });
      break;
    case 'bad-header-list':
      reply({ headers: { broken: [7] } });
      break;
    case 'bad-cookies':
      reply({ cookies: {} });
      break;
    case 'bad-cookie-item':
      reply({ cookies: [7] });
      break;
    case 'trailing-lines':
      reply();
      setTimeout(() => {
        process.stdout.write('not json at all\n');
        process.stdout.write(JSON.stringify({ status: 200, body: 'orphan' }) + '\n');
      }, 10);
      break;
    default:
      reply({ status: 404 });
  }
});
